numpy/core/code_generators/genapi.py

```diff
@@ -325,9 +328,19 @@ def array_api_define(self):
         return "        (void *) &%s" % self.name
 
     def internal_define(self):
-        astr = """\
-extern NPY_NO_EXPORT PyTypeObject %(type)s;
-""" % {'type': self.name}
+        if self.internal_type is None:
+            return f"extern NPY_NO_EXPORT {self.ptr_cast} {self.name};\n"
+
+        # If we are here, we need to define a larger struct internally, which
+        # the type can be cast safely. But we want to normally use the original
+        # type, so name mangle:
+        mangled_name = f"{self.name}Full"
+        astr = (
+            # Create the mangled name:
+            f"extern NPY_NO_EXPORT {self.internal_type} {mangled_name};\n"
+            # And define the name as: (*(type *)(&mangled_name))
+            f"#define {self.name} (*({self.ptr_cast} *)(&{mangled_name}))\n"
+        )
```

Comment:
The API generation changes are the ugliest part here... Unfortunately, I really want to hide everything away (unless it is an internal build), and have the `PyArrayDescr_Type` defined as `PyTypeObject` even internally. But I also need to initialize/allocate it statically as a `PyArray_DTypeMeta`.